ssl/ssl_sess.c

```diff
@@ -156,16 +156,12 @@ SSL_SESSION *SSL_get1_session(SSL *ssl)
 /* variant of SSL_get_session: caller really gets something */
 {
     SSL_SESSION *sess;
-    /*
-     * Need to lock this all up rather than just use CRYPTO_add so that
-     * somebody doesn't free ssl->session between when we check it's non-null
-     * and when we up the reference count.
-     */
-    CRYPTO_w_lock(CRYPTO_LOCK_SSL_SESSION);
+
     sess = ssl->session;
+    CRYPTO_MUTEX_lock_write(&sess->lock);
```

Comment:
@ghedo, thanks for clarifying. 

yes, as you said accessing ssl->session may be protected with ssl->lock. In SSL_set_session() also, it need to be serialised ssl->session access similarly.